crates/ide/src/references.rs

```diff
@@ -120,7 +120,13 @@ pub(crate) fn find_all_refs(
         }
         None => {
             let search = make_searcher(false);
-            Some(find_defs(sema, &syntax, position.offset)?.into_iter().map(search).collect())
+            Some(
+                find_defs(sema, &syntax, position.offset)?
+                    .into_iter()
+                    .unique()
+                    .map(search)
+                    .collect(),
+            )
```

Comment:
Hmm, this won't fix the linked issue though. In that issue, all the occurences are different `Definition`s! We already do deduplicate the results, its just the the resulting `NavigationTarget`s differ in properties that are not relevant to the shown results, so we need to do another deduplication pass when we convert to the requests response.

The deduplication should happen on the references before we return the `ReferenceSearchResult` here:
https://github.com/rust-lang/rust-analyzer/blob/5df53c96127095d75c0724f545ccd280dc0fafa9/crates/ide/src/references.rs#L106-L107

That is we need to go through all value `Vec`s of the map, sort, and deduplicate them there.

Reply:
> Hmm, this won't fix the linked issue though.

Good catch - that's because in that case the definitions are actually different, so unique() doesn't eliminate them, and because of the macro weirdness producing a cartesian product of references we also get redundant references along the technically-unique definitions.

> The deduplication should happen on the references before we return the ReferenceSearchResult here:

The results of the lambda created by `make_searcher` are already deduplicated. The problem is that we call said lambda multiple times and concatenate the results.